adds inline ability for es6 output

pflannery opened the pull request:
Adds a new transformer module called `InlineES6ModuleTransformer` for in-lining ES6 imports (as parts) with a root module.

usage: ```traceur --modules=inline --outputLanguage=es6 --out ./out/some-lib.js ./src/root.js```

transforms:

```js
 // a.js
 export function a() {
   return 'A';
 }

 // root.js
 import {a} from './a.js';
 export function b() {
   return a();
 }
```

into:

```js
 const $__module_a_js__ = (function() {
   "use strict";
   function a() {
     return 'A';
   }
   return {get a() {
       return a;
     }};
 })();
 "use strict";
 const {a} = $__module_a_js__;
 export function b() {
   return a();
 }
```

pflannery commented:
relates with #1842
blocked from running tests on windows because of https://github.com/google/traceur-compiler/pull/1832

arv commented:
Thanks!

Can this be tested? Maybe using a unit test.

pflannery commented:
@arv yes once I've figured out how to get the tests to run on windows. dont merge yet.

Is there a way to determine if the module being transformed is the top dominator? because when specifying a single source file in the cli it would be cool if it we could preserve expose it's es6 export keywords.

arv commented:
Is it `make` that is causing you trouble?

You can also run the tests directly so that you can skip the currently broken tests on windows.

pflannery commented:
its failing [here](https://github.com/google/traceur-compiler/pull/1832#issuecomment-87713400) and without the tval working it cant run any es6 test..

pflannery commented:
@arv I've added a test but not sure how to actually get the import (`test_a.js`, `test_b.js`) contents injected in to the result, any advice? 

and also before we merge this would be nice to be able to determine the root dominator file so we can preserve the ES6 export keywords for import to another ES6 library..

arv commented:
@pflannery Just to be clear. Is your use case that you want to compile a Module and its dependencies into a single ES6 Module. Something like:

Input:
```js
// a.js
export function a() {
  return 'A';
}
```
```js
// root.js
import {a} from './a.js';
export function b() {
  return a();
}
```
Outputs:
```js
// root-compiled.js
var $tmp = (function() {
  "use strict";
  function a() {
    return 'a';
  }
  return {a};
})();
const {a} = $tmp;
export function b() {
  return a();
}
```

pflannery commented:
@arv yes

pflannery commented:
@arv I've separated this transform in to a separate transformer module called `InlineES6ModuleTransformer` and updated the tests..

transforms:

```js
 // a.js
 export function a() {
   return 'A';
 }

 // root.js
 import {a} from './a.js';
 export function b() {
   return a();
 }
```

into:

```js
 const $__module_a_js__ = (function() {
   "use strict";
   function a() {
     return 'A';
   }
   return {get a() {
       return a;
     }};
 })();
 "use strict";
 var a = ($__module_a_js__).a;
 export function b() {
   return a();
 }
```

pflannery commented:
@arv PTAL

If good to go then I will rebase to current master and squash ;) thanks

arv commented on src/codegeneration/InlineES6ModuleTransformer.js:
```diff
@@ -0,0 +1,134 @@
+
+let anonInlineModules = 0;
+
+// transforms:
```
Make this into a single comment.

Can you also provide some information what the usage of this is?

arv commented on src/codegeneration/InlineES6ModuleTransformer.js:
```diff
@@ -0,0 +1,134 @@
+  // tests if we are a root module or not
+  isRootModule() {
+    let rootModules = this.options_.rootModules || [];
+    return rootModules.some((testModule)=> {
```
ws betwen `)` and `=>`

arv commented on src/codegeneration/InlineES6ModuleTransformer.js:
```diff
@@ -0,0 +1,134 @@
+    return createBindingIdentifier(this.getTempVarNameForModuleSpecifier(tree));
+  }
+
+  // tests if we are a root module or not
```
You can skip this comment. The method name provides the same information

arv commented on src/codegeneration/InlineES6ModuleTransformer.js:
```diff
@@ -0,0 +1,134 @@
+    return '$__module__' +
+      moduleName
+      .split(/[^a-zA-Z0-9$]/g)
+      .filter((value) => {return value;})
```
.filter((value) => value})

arv commented on src/codegeneration/PureES6Transformer.js:
```diff
@@ -63,5 +64,11 @@ export class PureES6Transformer extends MultiTransformer {
     append(AnnotationsTransformer);
     append(TypeTransformer);
+
+    // inline if the option was specified
```
remove comment

pflannery replied:
will do this and reformat other comments to use doc comments `/**`

arv commented on src/node/recursiveModuleCompile.js:
```diff
@@ -50,6 +50,8 @@ function recursiveModuleCompileToSingleFile(outputFile, includes, options) {
     return include;
   });
 
+  options.rootModules = resolvedIncludes;
```
I wonder if we could pass this information in some other way? It feels a bit odd to put this on the options object.

pflannery replied:
Maybe via metadata but it will need to be passed to the transformer too. There is a fair amount of complex workflow to get there..

arv replied:
@johnjbarton Do you have any suggestion here?

johnjbarton replied:
My initial reaction was to agree with @pflannery: the better solution is metadata and it is probably better to do that in a separate PR.  Then #1859

pflannery replied:
:+1:  thanks @johnjbarton

arv commented on test/unit/codegeneration/PureES6Transformer.js:
```diff
@@ -48,4 +48,44 @@ suite('PureES6Transformer.js', function() {
   });
+
+  test('Inline', function() {
+    var options = new Options({
```
Why the getters?

```js
var options = new Options({
  modules: 'inline',
  rootModules: {name: null}
})
```

arv commented on test/unit/codegeneration/PureES6Transformer.js:
```diff
@@ -48,4 +48,44 @@ suite('PureES6Transformer.js', function() {
+    var reporter = new ErrorReporter();
+
+    var code = [
+      "import {TestA} from './resources/test_a.js';",
```
We use single quotes everywhere... but in this case you can use a template literal.

arv commented on src/codegeneration/InlineES6ModuleTransformer.js:
```diff
@@ -0,0 +1,134 @@
+    });
+  }
+
+  getTempVarNameForModuleName(moduleName) {
```
It is not clear why we cannot use the existing logic here?

pflannery replied:
will remove

arv commented on src/codegeneration/InlineES6ModuleTransformer.js:
```diff
@@ -0,0 +1,134 @@
+
+  // override the module entry point to place "use strict" at the top
+  moduleProlog() {
+    return [createUseStrictDirective()];
```
Modules are always strict so adding this is just noise.

pflannery replied:
I should be a bit clearer here, it prevents the default ```var __moduleName = "../some/path/to/module.js";``` being generated and for every module

arv replied:
Maybe just return `[]`?

pflannery replied:
just trying this and it removes 'use strict' all together

arv replied:
Isn't that what we want?

pflannery replied:
if that is the case then would it need to put one "use strict" at the top of the output or does "use strict" hoist?

arv replied:
The output of this transformer is a Module which is already strict.

arv commented on src/codegeneration/InlineES6ModuleTransformer.js:
```diff
@@ -0,0 +1,134 @@
+
+  // override to wrap module in a closure unless it's a root module
+  wrapModule(statements) {
+    let seed = this.moduleName || 'anon_' + ++anonInlineModules;
```
It is not clear to me when we do not have a module name here?

pflannery replied:
me too, I've just reused the existing inline transformer logic [from here](https://github.com/google/traceur-compiler/blob/master/src/codegeneration/InlineModuleTransformer.js#L38)

johnjbarton replied:
I guess only roots of the dependency tree can be anonymous, eg the ones coming via `Loader.module()`.

arv commented:
Very nice.

Can you look at the Travis error too?

pflannery commented:
@arv @johnjbarton I've implemented root module detection using metadata and removed code that references the options object. 

There is a case where the metadata is not defined when using the --dir switch, however I'm not sure anyone would ever use '--dir', `inline` and `es6`  together but it does work. Because of this I've made sure that the metadata object is defined when evaluating `isRootModule`.

PTAL

johnjbarton commented:
Thanks, the metadata approach looks much better.

I'm ok with this but for your sake I think another PR with a test with the command line (generated-code-dependencies.js) would be a good addition. The test would take two modules, generate output and run the resulting file; the root module would just use a value from its dependency. This exercises all of the elements without being effected by the details of the code gen.  Especially in this case the rootModule which is null in the current test.

pflannery commented:
@johnjbarton no problem. I can add the tests to this PR.

pflannery commented:
Glad I did this as I found an issue where I needed to replicate what `InlineModuleTransformer` does with `transformNamedExport`
```js
transformNamedExport(tree) {
    return createEmptyStatement();
}
```
There seems to be a little bug with this as an empty line with `;` get generated,  I see this happening with the current InlineModuleTransformer too. 
It doesn't break the tests and is valid runnable code. I believe this also relates to #1822

arv commented:
@pflannery You can return an AnonBlock instead. Those gets removed/expanded in transformList.

arv commented on src/codegeneration/PureES6Transformer.js:
```diff
@@ -65,5 +66,9 @@ export class PureES6Transformer extends MultiTransformer {
     append(AnnotationsTransformer);
     append(TypeTransformer);
+
+    if (options.modules.toLowerCase() === 'inline') {
```
I would skip the toLowerCase here. No need for us to support case insensitive values here.

pflannery replied:
ok I will update this now

arv commented:
LGTM

Thanks for taking care of this.

pflannery commented:
> You can return an AnonBlock instead. Those gets removed/expanded in transformList.

@arv I've switched to use 
```js
transformNamedExport(tree) {
    return new AnonBlock(null, []);
}
```

arv commented:
Actually, I still don't understand why you need to add the `'use strict'` to the top of your module?

pflannery commented:
@arv currently with iojs and chrome canary with `--harmony-classes` enabled I'm seeing this error
```js
SyntaxError: Block-scoped declarations (let, const, function, class) not yet supported outside strict mode
```

pflannery commented:
if there isn't a way around this then I could put a `TODO` to remove this once this condition is lifted?

arv commented:
Lets leave the use strict for now. It does not hurt.

However, it is not clear what the intended output is? Right now ModuleTransformer transforms a Module into a Script and a Script cannot contain an ExportDeclaration (validation is always off for PureES6Transformer). But what we really want to do is to transform a Module into a Module and execute it as a module, which of course is not possible with V8 today.

arv commented:
Merging
